R/trackhub.R

```diff
@@ -6,64 +6,123 @@
+    recordsList
 }
 
-hubFile <- function(x) file.path(uri(x), "hub.txt")
+isFileReference <- function(x)    {
+  tools::file_ext(x) == "txt"
+}
 
+hubFile <- function(x) file.path(uri(x), "hub.txt")
 genomesFile <- function(x, y) file.path(uri(x), y)
 
-setMethod("genome", "TrackHub", function(x) {
-  hubContent <- getContent(hubFile(x))
-  field <- "genomesFile"
-  genomesFileValue <- getValue(field, hubContent)
-  if (isFileReference(genomesFileValue)) {
-    path <- genomesFile(x, genomesFileValue)
-    genomeContent <- getContent(path)
-    genomes <- getValue("genome", genomeContent)
+setMethod("genome", "TrackHub", function(x)    {
+  hubContent <- getHubContent(hubFile(x))
+  genomesFileValue <- hubContent["genomesFile"]
+  if ((isFileReference(genomesFileValue) && !is.na(genomesFileValue)))    {
+    genomesFilePath <- genomesFile(x, genomesFileValue)
+    genomesRecordsList <- getGenomesContent(genomesFilePath)
+    genomes <- sapply(genomesRecordsList, function(x) x["genome"])
+    as.character(genomes)
   }
-  else genomes <- genomesFileValue
-  as.character(structure(genomes, name = field))
+  else stop("hub.txt: 'genomesFile' does not contain valid reference to genomes file")
+})
+
+setMethod("[[", "TrackHub", function (x, i, j, ...)    {
+  if (!missing(j))
+    warning("argument 'j' ignored")
+  TrackHubGenome(x, i, ...)
 })
 
-setMethod("length", "TrackHub", function(x) length(genome(x)))
+setMethod("$", "TrackHub", function (x, name)    {
+  TrackHubGenome(x, name)
+})
+
+setMethod("names", "TrackHub", function(x) genome(x))
 
-setMethod("show", "TrackHub", function(object) {
+setMethod("length", "TrackHub", function(x) length(names(x)))
+
+setMethod("show", "TrackHub", function(object)    {
   cat(class(object), "repository\nuri:", uri(object), "\n")
   cat(S4Vectors:::labeledLine("genomes", genome(object)))
 })
 
-TrackHub <- function(uri="TrackHub", create = FALSE){
-   if (!isTRUEorFALSE(create))
+TrackHub <- function(uri, create = FALSE)    {
+  if (!isTRUEorFALSE(create))
     stop("'create' must be TRUE or FALSE")
-  if (create) {
-    if (uriExists(uri)) {
+  if (create)    {
+    if (uriExists(uri))    {
       message("NOTE: '", uri, "' already exists")
       create <- FALSE
     } ## must create this before calling normURI (requires existence)
     else createResource(uri, dir = TRUE)
   }
-  ql <- new("TrackHub", uri = normURI(uri))
+  th <- new("TrackHub", uri = normURI(uri))
   if (create)
-    createResource(hubFile(ql))
-  ql
+    createResource(hubFile(th))
+  th
 }
 
 setAs("character", "TrackHub", function(from) TrackHub(from))
+
+### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+### TrackHubGenome class
+###
+
+setClass("TrackHubGenome",
+         representation(trackhub = "TrackHub",
+                        genome = "character"),
+         contains = "TrackDb")
+
+trackhub <- function(x, ...) x@trackhub
+trackDbFile <- function(x,y) file.path(uri(x), y)
```

Comment:
Understood, Yes this can cause problems. Maybe we can create a new function that will check URI is local or not with the help of `uriIsLocal` function. If it's local we can use `file.path` otherwise, URI can be combined with `paste0` function.
```R
uriConcat <- function(x,y) {
    uri <- .parseURI(uri(x))
    if (uriIsLocal(uri)) {
      file.path(uri(x), y)
    }else {
      paste(uri(x), y, sep="/")
    }
}
```

Or is there any better solution?

Reply:
Why not always use `paste()` since the URL spec requires paths to use the `/` separator?

Reply:
yes just ` paste()` is sufficient for this purpose.
I thought of handling URI and URL separately and in doing so I miss that both of them can be solved with `paste()`.